Update ufs-bundle to build with head of develop of ufs-weather-model for ATM configuration

mark-a-potts opened the pull request:
## Description

This PR, in combination with PRs for oops and fv3-jedi, allows the ufs-bundle to be built using the head of develop for the ufs-weather-model.  Additionally, new regression tests have been added to fv3-jedi that will update the run directories needed for the UFS model and to run a sample Ensemble Forecast of the UFS. 

## Issue(s) addressed

Resolves #14 

## Dependencies

List the other PRs that this PR is dependent on:
- [x] https://github.com/JCSDA-internal/fv3-jedi/pull/1099


## Impact

None. 

## Checklist

- [X] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [X] I have run the unit tests before creating the PR

mark-a-potts commented:
This PR has been tested and works with UFS_APP=ATM, ATMAERO, and S2S configurations.  There are no tests for ATMAERO or S2S, but both configurations build. The NG-GODAS option does not build because the soca branch required is out of date with the rest of the bundle and requires fairly significant work.

climbfuji commented:
> This PR has been tested and works with UFS_APP=ATM, ATMAERO, and S2S configurations. There are no tests for ATMAERO or S2S, but both configurations build. The NG-GODAS option does not build because the soca branch required is out of date with the rest of the bundle and requires fairly significant work.

I have an updated SOCA branch that I used for a similar effort ... I can make that work with this PR.

mark-a-potts commented:
That would be great.  I started on a merge and it looked like it was going to be a lot of work. I added a patch to handle an issue with checksum checking of restart files down in atmos_cubed_sphere. I know that there are differences in the MOM6 repo, so maybe we could add a similar patch for those and get everything working with the head of develop of the WM.

climbfuji commented:
@mark-a-potts Let's fix the CI failures first. This is because of a merge last week on the JEDI side that changed the model interfaces, but nobody thought about the fv3-jedi etc branches for the UFS. Let me do this in an hour or so, ok?

mark-a-potts commented:
I have updated this to now handle building ATM, ATMAERO, and S2S using the head of develop of the ufs-wm. NG-GODAS is still very broken, but the build with S2S will actually run the ATM regression tests without issue.

climbfuji commented:
@mark-a-potts I resolved the merge conflict in https://github.com/JCSDA/ufs-bundle/pull/43/commits/dd72818159051e4e0958ef0d9a6ce9df78b1ff43, please take a look to see if everything looks ok

climbfuji commented on README.md:
```diff
@@ -4,7 +4,7 @@ Bundle for interfacing UFS models with JEDI interfaces
 ## Required thirdparty libraries
 
-This bundle requires the following spack-stack modules be loaded (all except `fms@2022.04` are in the `skylab-3.0.0` environment):
+This bundle requires the following spack-stack modules be loaded (all except `fms@2022.04` and `atlast@0.35.1` are in the `skylab-3.0.0` environment):
```
We need to update this section (and the list of modules) to something more recent

climbfuji commented:
There is now a CI failure in the cmake step, related to fv3-jedi-data: https://github.com/JCSDA/ufs-bundle/actions/runs/7133595917/job/19426615092?pr=43

mark-a-potts commented:
Just removed the extra fv3-jedi-data section. I think it should work now.

climbfuji commented:
Sigh, there are so many updates to the JEDI code that it seems impossible to get a clean run! I'll hop on the instance and clear the build area, then restart the test.

mark-a-potts commented:
The latest update to ioda seems to have broken the buid now.

climbfuji commented:
> The latest update to ioda seems to have broken the buid now.

Can you try to pull oops develop into your branch please?

mark-a-potts commented:
Yeah, I think that was the issue. Just re-built successfully on my machine, so pushing now.

climbfuji commented on CMakeLists.txt:
```diff
@@ -168,6 +175,28 @@ else()
+
+# When LOCAL_PATH_JEDI_TESTFILES is set to the directory of IODA test files stored
+# in a local directory, ioda-data repo will not be cloned
+if( NOT DEFINED ENV{LOCAL_PATH_JEDI_TESTFILES} )
```
Note that this is not how we want to handle those switches in the future. Instead, the logic in the bundles should be like here: https://github.com/JCSDA-internal/jedi-bundle/blob/551edcbc73700544994c2b10fa1270fa25c1900a/CMakeLists.txt#L51C1-L51C91

climbfuji commented on CMakeLists.txt:
```diff
@@ -95,12 +89,26 @@ else()
+  set(UFS_REPO https://github.com/ufs-community/ufs-weather-model)
+  set(UFS_TAG "develop" )
+  set(patch_command ${CMAKE_COMMAND} -E copy_if_different ${CMAKE_CURRENT_SOURCE_DIR}/patches/mom6.patch <SOURCE_DIR>/MOM6-interface && cd <SOURCE_DIR>/MOM6-interface && git apply mom6.patch)
+else(UFS_APP MATCHES "^(NG-GODAS)$" )
```
Is this still needed, given that we updated the soca branch?

mark-a-potts replied:
If we switch to the branch I used for the PR into the WM, we can probably combine all APPs together, though there will still be a patch required to get ng-godas to build. Since the patches tend to be kind of fragile though, it might be best to only apply the patch for UFS_APP=NG-GODAS.

climbfuji commented:
@mark-a-potts I am also running into problems when having to rerun make. Whether UPDATE_DISCONNECTED is ON or OFF, I get something like this:
```
[  0%] Performing update step for 'ufs-weather-model'
cd /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/ufs-weather-model && /opt/homebrew/Cellar/cmake/3.27.0/bin/cmake -Dcan_fetch=YES -P /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake
No stash entries found.
No stash entries found.
No stash entries found.
CMake Error at /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake:277 (message):


  Failed to unstash changes in:
  '/Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/ufs-weather-model'.


  You will have to resolve the conflicts manually


make[2]: *** [ufs-weather-model/src/ufs-weather-model-stamp/ufs-weather-model-update] Error 1
make[1]: *** [CMakeFiles/ufs-weather-model.dir/all] Error 2
make: *** [all] Error 2
```
This may have to do with the patch that gets applied to the ufs-weather-model develop branch?

climbfuji commented:
> @mark-a-potts I am also running into problems when having to rerun make. Whether UPDATE_DISCONNECTED is ON or OFF, I get something like this:
> 
> ```
> [  0%] Performing update step for 'ufs-weather-model'
> cd /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/ufs-weather-model && /opt/homebrew/Cellar/cmake/3.27.0/bin/cmake -Dcan_fetch=YES -P /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake
> No stash entries found.
> No stash entries found.
> No stash entries found.
> CMake Error at /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake:277 (message):
> 
> 
>   Failed to unstash changes in:
>   '/Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/ufs-weather-model'.
> 
> 
>   You will have to resolve the conflicts manually
> 
> 
> make[2]: *** [ufs-weather-model/src/ufs-weather-model-stamp/ufs-weather-model-update] Error 1
> make[1]: *** [CMakeFiles/ufs-weather-model.dir/all] Error 2
> make: *** [all] Error 2
> ```
> 
> This may have to do with the patch that gets applied to the ufs-weather-model develop branch?

Update. I did a clean checkout and ran into the same problem:
```
[  0%] Performing disconnected update step for 'ufs-weather-model'
cd /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/ufs-weather-model && /opt/homebrew/Cellar/cmake/3.27.0/bin/cmake -Dcan_fetch=NO -P /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake
No stash entries found.
No stash entries found.
No stash entries found.
CMake Error at /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake:277 (message):


  Failed to unstash changes in:
  '/Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ensforecasts/ufs-weather-model'.


  You will have to resolve the conflicts manually


make[2]: *** [ufs-weather-model/src/ufs-weather-model-stamp/ufs-weather-model-update_disconnected] Error 1
make[1]: *** [CMakeFiles/ufs-weather-model.dir/all] Error 2
make: *** [all] Error 2
```
Since the patch only applies to the top-level ufs-weather-model and not to any of the submodules underneath, I would say we use a branch for the moment. The ufs-weather-model developers have indicated that they are open to these changes, therefore I expect that we wouldn't have to maintain that ufs-weather-model branch for very long.

mark-a-potts commented:
I cleaned up the CMakeLists.txt file a bit and fixed the *-data bundle syntax to match the other bundles. Note that the ufs-weather-model is now pointing to my PR branch (feature/fortran-flags) in github.com/NOAA-EPIC/ufs-weather-model. I got rid of the mom6 patch and added an ng-godas patch instead. It gets run at the top level of ufs-weather-model, but it is actually patching the MOM6 repo under MOM6-interface.  I have successfully built using UFS_APP=NG-GODAS and S2S, but still waiting on ATM and ATMAERO.

climbfuji commented:
> I cleaned up the CMakeLists.txt file a bit and fixed the *-data bundle syntax to match the other bundles. Note that the ufs-weather-model is now pointing to my PR branch (feature/fortran-flags) in github.com/NOAA-EPIC/ufs-weather-model. I got rid of the mom6 patch and added an ng-godas patch instead. It gets run at the top level of ufs-weather-model, but it is actually patching the MOM6 repo under MOM6-interface. I have successfully built using UFS_APP=NG-GODAS and S2S, but still waiting on ATM and ATMAERO.

That sounds great, thanks very much @mark-a-potts

climbfuji commented:
@mark-a-potts Looks like CI tests are passing now !

mark-a-potts commented:
They didn't actually all run @climbfuji  But, it is close now. I will try submitting again.

mark-a-potts commented:
This time the tests actually ran and seem to have passed. Yay!

climbfuji commented on CMakeLists.txt:
```diff
@@ -132,42 +145,41 @@ add_library( stochastic_physics IMPORTED STATIC)
+  ecbuild_bundle( PROJECT fv3-jedi GIT "https://github.com/jcsda-internal/fv3-jedi.git" BRANCH feature/ufs-EnsForecasts UPDATE )  
 elseif(UFS_APP MATCHES "^(NG-GODAS)$")
-  ecbuild_bundle( PROJECT soca     GIT "https://github.com/jcsda-internal/soca.git" BRANCH feature/ufs_dom UPDATE )  # updated from develop Dec 12 2023
+  ecbuild_bundle( PROJECT soca     GIT "https://github.com/jcsda-internal/soca.git" BRANCH feature/ufs_dom )  
```
I think we should keep the `UPDATE` here as well?

mark-a-potts replied:
The soca branch needs some updates as well. I will put in a PR to feature/ufs_dom

climbfuji reviewed (commented):
I think this PR is ready for review, isn't it?

climbfuji commented:
@mark-a-potts I am unable to build any of the APPs with this PR. NO matter what I do, I get this from the `make` command after a fresh checkout and fresh cmake run:
```
[  0%] Performing disconnected update step for 'ufs-weather-model'
cd /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/ufs-weather-model && /opt/homebrew/Cellar/cmake/3.27.0/bin/cmake -Dcan_fetch=NO -P /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake
No stash entries found.
No stash entries found.
No stash entries found.
CMake Error at /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake:277 (message):


  Failed to unstash changes in:
  '/Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/ufs-weather-model'.


  You will have to resolve the conflicts manually


make[2]: *** [ufs-weather-model/src/ufs-weather-model-stamp/ufs-weather-model-update_disconnected] Error 1
make[1]: *** [CMakeFiles/ufs-weather-model.dir/all] Error 2
make: *** [all] Error 2
```

mark-a-potts commented:
> @mark-a-potts I am unable to build any of the APPs with this PR. NO matter what I do, I get this from the `make` command after a fresh checkout and fresh cmake run:
> 
> ```
> [  0%] Performing disconnected update step for 'ufs-weather-model'
> cd /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/ufs-weather-model && /opt/homebrew/Cellar/cmake/3.27.0/bin/cmake -Dcan_fetch=NO -P /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake
> No stash entries found.
> No stash entries found.
> No stash entries found.
> CMake Error at /Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/build-atm/ufs-weather-model/tmp/ufs-weather-model-gitupdate.cmake:277 (message):
> 
> 
>   Failed to unstash changes in:
>   '/Users/heinzell/work/ufs-bundle/ufs-bundle-mark-ufs-ensforcasts/ufs-weather-model'.
> 
> 
>   You will have to resolve the conflicts manually
> 
> 
> make[2]: *** [ufs-weather-model/src/ufs-weather-model-stamp/ufs-weather-model-update_disconnected] Error 1
> make[1]: *** [CMakeFiles/ufs-weather-model.dir/all] Error 2
> make: *** [all] Error 2
> ```

So, this might be because you need to remove the old ufs-weather-model directory? To accommodate the addition of the fortran flags for building, I switched to a fork of the WM under NOAA-EPIC, so if it still thinks it is looking for the branch under ufs-community, it will fail.

mark-a-potts commented:
I just checked with @jkbk2004, and it sounds like my PR to the UFS-WM should go in later this week, but in the meantime, I extended the patch and will push up a new CMakeLists.txt file that pulls from ufs-community instead of NOAA-EPIC.

climbfuji reviewed (changes requested):
I know why I get this failure even for a completely fresh checkout. It's because of this: https://github.com/NOAA-EMC/AQM/pull/40

So, long story short, the patch logic needs to be removed or otherwise we can't build/test on macOS any longer.

mark-a-potts commented:
If it is exclusively a case-sensitivity issue, this might be an alternative for your testing--

https://brianboyko.medium.com/a-case-sensitive-src-folder-for-mac-programmers-176cc82a3830

We should be able to remove the patch to fix the CMakeLists.txt files when my PR into the WM gets merged, but the rest of that patch fixes issues in MOM6 that we need for coupling to SOCA. Given that the NG-GODAS coupling is broken right now anyway, I would probably lean towards removing the NG-GODAS option from ufs-bundle for the time being.  

We could also put in a PR to MOM6, but I feel kind of uncomfortable doing that if the SOCA connection doesn't really work right now.

climbfuji commented:
I agree, let's hold off with the PR to mom6 until we know it is working. I was able to build it on my macOS using the case-sensitive volume trick you mentioned above. That's not ideal, but I don't see any better option either. Definitely we need to document this in the top-level `README.md` of this repository (i.e in this PR).

climbfuji reviewed (commented):
This works for me if I am using a case-sensitive volume on my macOS. We need to get the dependent PRs approved before we can approve and merge this.

mark-a-potts commented:
The oops PR is going to take a while, but I think I can comment out the ensemble test in fv3-jedi and get this working. I'll try that today.

climbfuji commented on CMakeLists.txt:
```diff
@@ -53,12 +55,13 @@ endif()
 option(ENABLE_IODA_DATA "Obtain ioda test data from ioda-data repository (vs tarball)" ON)
-ecbuild_bundle( PROJECT ioda-data GIT "https://github.com/JCSDA-internal/ioda-data.git" BRANCH develop UPDATE )
 ecbuild_bundle( PROJECT ioda      GIT "https://github.com/jcsda-internal/ioda.git"      BRANCH develop UPDATE )
+ecbuild_bundle( PROJECT ioda-data GIT "https://github.com/JCSDA-internal/ioda-data.git" BRANCH develop UPDATE )
```
Why does this need to be moved from above ioda to below ioda?

mark-a-potts replied:
Fixed this.

climbfuji commented on CMakeLists.txt:
```diff
@@ -53,12 +55,13 @@ endif()
 option(ENABLE_UFO_DATA "Obtain ufo test data from ufo-data repository (vs tarball)" ON)
-ecbuild_bundle( PROJECT ufo-data GIT "https://github.com/jcsda-internal/ufo-data.git" BRANCH develop UPDATE )
-ecbuild_bundle( PROJECT ufo      GIT "https://github.com/jcsda-internal/ufo.git"      BRANCH develop UPDATE )
+ecbuild_bundle( PROJECT ufo-data GIT "https://github.com/JCSDA-internal/ufo-data.git"   BRANCH develop UPDATE )
```
We should be able to revert the changes in line 60 and 61

mark-a-potts replied:
I think this is reverted now, too.

climbfuji commented on README.md:
```diff
@@ -4,7 +4,7 @@ Bundle for interfacing UFS models with JEDI interfaces
 
```
Please also provide some background information and the link for creating the case-sensitive volume on macOS to compile w/ the patch logic.

mark-a-potts replied:
Done.

climbfuji reviewed (commented):
We are getting there, thanks for all your work!

climbfuji reviewed (dismissed):
Thanks very much for the `README.md` updates and your patience along the way!

climbfuji commented:
Last thing is to get the fv3-jedi PR merged, correct? I think you avoided the oops dependency (it's still in the PR description) for now.

mark-a-potts commented:
Yes. The fv3-jedi PR can be done without the oops PR, and I think it would be good to do that. I will remove the mentions of oops in the PR descriptions.

climbfuji commented on CMakeLists.txt:
```diff
@@ -132,42 +146,42 @@ add_library( stochastic_physics IMPORTED STATIC)
-    ecbuild_bundle( PROJECT soca        GIT "https://github.com/jcsda-internal/soca.git"          BRANCH feature/ufs_dom UPDATE )  # updated from develop Dec 12 2023
-    add_dependencies(soca ufs-weather-model)
-  endif()
+  ecbuild_bundle( PROJECT fv3-jedi-data GIT "https://github.com/JCSDA-internal/fv3-jedi-data.git" BRANCH feature/ufs-EnsForecasts ) 
```
I still see a branch `feature/ufs-EnsForecasts` for fv3-jedi-data ?

climbfuji commented on CMakeLists.txt:
```diff
@@ -132,42 +146,42 @@ add_library( stochastic_physics IMPORTED STATIC)
-    add_dependencies(soca ufs-weather-model)
-  endif()
+  ecbuild_bundle( PROJECT fv3-jedi-data GIT "https://github.com/JCSDA-internal/fv3-jedi-data.git" BRANCH feature/ufs-EnsForecasts ) 
+  ecbuild_bundle( PROJECT fv3-jedi GIT "https://github.com/jcsda-internal/fv3-jedi.git" BRANCH feature/ufs-EnsForecasts UPDATE )  
```
The fv3-jedi branch `feature/ufs-EnsForecasts` was just merged into `develop`

climbfuji commented on CMakeLists.txt:
```diff
@@ -132,42 +146,42 @@ add_library( stochastic_physics IMPORTED STATIC)
 elseif(UFS_APP MATCHES "^(NG-GODAS)$")
-  ecbuild_bundle( PROJECT soca     GIT "https://github.com/jcsda-internal/soca.git" BRANCH feature/ufs_dom UPDATE )  # updated from develop Dec 12 2023
+# ecbuild_bundle( PROJECT soca     GIT "https://github.com/jcsda-internal/soca.git" BRANCH feature/ufs_dom UPDATE )  
+  ecbuild_bundle( PROJECT soca     GIT "https://github.com/jcsda-internal/soca.git" BRANCH feature/ufs_mark UPDATE )  
```
Has this been merged already?

climbfuji reviewed (commented):
Please update the PR from ufs-bundle develop, revert the fv3-jedi branch name; this PR also uses temporary branches for soca and fv3-jedi-data?

climbfuji commented:
There is also a conflict for `README.md` because my last PR replaced `w3nco` with `w3emc`.

mark-a-potts commented:
Working on this now. The s3 bucket is currently being synched to the latest rt results, so that is delaying the testing a bit. Will update shortly.

mark-a-potts commented:
After the latest update, the ctests pass and builds now work for ATM, ATMAERO, NG-GODAS, and S2S configurations.

climbfuji commented on .github/workflows/run_ec2_pcluster.yaml:
```diff
@@ -135,38 +136,35 @@ jobs:
           cd ${JEDI_ENV}
           mkdir -p build-atmaero
           cd build-atmaero
+          rm -rf *
```
Why is this needed?

mark-a-potts replied:
I had to put it there to make sure the build was started fresh. I'll remove it now.

climbfuji commented on CMakeLists.txt:
```diff
@@ -58,9 +60,10 @@ ecbuild_bundle( PROJECT ioda-data GIT "https://github.com/JCSDA-internal/ioda-da
 
 option(ENABLE_UFO_DATA "Obtain ufo test data from ufo-data repository (vs tarball)" ON)
-ecbuild_bundle( PROJECT ufo-data GIT "https://github.com/jcsda-internal/ufo-data.git" BRANCH develop UPDATE )
+ecbuild_bundle( PROJECT ufo-data GIT "https://github.com/JCSDA-internal/ufo-data.git" BRANCH develop UPDATE )
```
```suggestion
ecbuild_bundle( PROJECT ufo-data GIT "https://github.com/jcsda-internal/ufo-data.git" BRANCH develop UPDATE )
```

climbfuji commented on CMakeLists.txt:
```diff
@@ -58,9 +60,10 @@ ecbuild_bundle( PROJECT ioda-data GIT "https://github.com/JCSDA-internal/ioda-da
+ecbuild_bundle( PROJECT ufo-data GIT "https://github.com/JCSDA-internal/ufo-data.git" BRANCH develop UPDATE )
 ecbuild_bundle( PROJECT ufo      GIT "https://github.com/jcsda-internal/ufo.git"      BRANCH develop UPDATE )
 
+
```
```suggestion
```

climbfuji commented on CMakeLists.txt:
```diff
@@ -133,42 +147,41 @@ add_library( stochastic_physics IMPORTED STATIC)
   message(FATAL_ERROR "ufs-bundle unknown UFS_APP ${UFS_APP}")
 endif()
 
+
```
```suggestion
```

climbfuji commented on README.md:
```diff
@@ -4,34 +4,41 @@ Bundle for interfacing UFS models with JEDI interfaces
+ 21) libpng/1.6.37                    47) ectrans/1.2.0     73) py-f90nml/1.4.3              99) nemsio/2.5.4
+ 22) g2/3.4.5                         48) atlas/0.35.0      74) py-h5py/3.7.0               100) sfcio/1.4.1
+ 23) g2tmpl/1.10.2                    49) git-lfs/3.0.2     75) py-cftime/1.0.3.4           101) sigio/2.3.2
+ 24) sp/2.3.3                         50) gsibec/1.1.3      76) py-netcdf4/1.5.8            102) w3nco/2.4.1
```
This is reverting my recent change in `README.md` that replaced `w3nco` with `w3emc`

mark-a-potts replied:
Ah, I missed that. I mostly wanted to update the fms version to 2023.02.01 which is now required by the WM. I'll get that straightened out.

climbfuji reviewed (commented):
A few minor comments to reduce the amount of unnecessary changes etc, other than that this looks great!

climbfuji reviewed (approved):
Thanks for all your work on this!

climbfuji commented:
@mark-a-potts CI tests are failing unfortunately: https://github.com/JCSDA/ufs-bundle/actions/runs/7745452671/job/21121475447?pr=43

climbfuji reviewed (approved):
Would have been nice to keep the build directories (much faster), but if it doesn't work ...
